File: scratch/count_components_tokens.js
const fs = require('fs');
const path = require('path');

function walk(dir) {
  let results = [];
  const list = fs.readdirSync(dir);
  list.forEach(file => {
    if (file.startsWith('.')) return;
    const filePath = path.join(dir, file);
    const stat = fs.statSync(filePath);
    if (stat && stat.isDirectory()) {
      results = results.concat(walk(filePath));
    } else {
      results.push(filePath);
    }
  });
  return results;
}

function estimateTokens(text) {
  // A good estimate for code (which has many spaces, brackets, short keywords)
  // is around 3.2 to 3.5 characters per token. 
  // Let's provide a range or a reasonable average like 3.3.
  if (!text) return 0;
  return Math.ceil(text.length / 3.3);
}

function main() {
  const targetDir = path.resolve(__dirname, '..', 'components');
  if (!fs.existsSync(targetDir)) {
    console.error(`Error: Directory ${targetDir} does not exist.`);
    process.exit(1);
  }

  console.log(`Analyzing directory: ${targetDir}`);
  console.log('--------------------------------------------------------------------------------');

  const files = walk(targetDir);
  const fileDetails = [];
  let totalChars = 0;
  let totalTokens = 0;
  let totalLines = 0;

  files.forEach(file => {
    try {
      const content = fs.readFileSync(file, 'utf8');
      const chars = content.length;
      const tokens = estimateTokens(content);
      const lines = content.split('\n').length;
      const relPath = path.relative(targetDir, file);

      fileDetails.push({
        relPath,
        chars,
        tokens,
        lines
      });

      totalChars += chars;
      totalTokens += tokens;
      totalLines += lines;
    } catch (e) {
      console.error(`Error reading file ${file}:`, e.message);
    }
  });

  // Sort files by estimated tokens descending
  fileDetails.sort((a, b) => b.tokens - a.tokens);

  console.log(`${'File Path'.padEnd(55)} | ${'Lines'.padStart(6)} | ${'Characters'.padStart(10)} | ${'Est. Tokens'.padStart(12)}`);
  console.log('-'.repeat(91));
  fileDetails.forEach(f => {
    console.log(`${f.relPath.replace(/\\/g, '/').padEnd(55)} | ${f.lines.toString().padStart(6)} | ${f.chars.toLocaleString().padStart(10)} | ${f.tokens.toLocaleString().padStart(12)}`);
  });
  console.log('-'.repeat(91));
  console.log(`Total Files:         ${fileDetails.length}`);
  console.log(`Total Lines:         ${totalLines.toLocaleString()}`);
  console.log(`Total Characters:    ${totalChars.toLocaleString()}`);
  console.log(`Estimated Tokens:    ${totalTokens.toLocaleString()} (approx 3.3 chars/token)`);
  console.log(`Estimated (4 chars):  ${Math.ceil(totalChars / 4).toLocaleString()} tokens`);
}

main();

File: scratch/count_components_tokens.py
import os
import sys

# Try to import tiktoken
try:
    import tiktoken
    has_tiktoken = True
except ImportError:
    has_tiktoken = False

def count_tokens(text, encoding_name="cl100k_base"):
    if has_tiktoken:
        try:
            encoder = tiktoken.get_encoding(encoding_name)
            return len(encoder.encode(text, disallowed_special=()))
        except Exception as e:
            pass
    # Fallback to characters // 4
    return max(1, len(text) // 4) if text else 0

def main():
    target_dir = r"c:\Users\david\flashcheckout\components"
    if not os.path.exists(target_dir):
        print(f"Error: Directory {target_dir} does not exist.")
        sys.exit(1)
        
    print(f"Analyzing tokens in: {target_dir}")
    print(f"Using tiktoken: {has_tiktoken}")
    if has_tiktoken:
        print("Note: Using cl100k_base encoding.")
    else:
        print("Note: tiktoken not found. Using fallback of approx. 4 characters per token.")
    print("-" * 80)
    
    total_files = 0
    total_chars = 0
    total_tokens = 0
    
    # We want to list details of files sorted by token count
    file_details = []
    
    for root, dirs, files in os.walk(target_dir):
        # Skip hidden directories
        dirs[:] = [d for d in dirs if not d.startswith('.')]
        
        for file in files:
            if file.startswith('.'):
                continue
            file_path = os.path.join(root, file)
            rel_path = os.path.relpath(file_path, target_dir)
            
            try:
                with open(file_path, 'r', encoding='utf-8', errors='replace') as f:
                    content = f.read()
            except Exception as e:
                print(f"Skipping {rel_path} due to error: {e}")
                continue
                
            chars = len(content)
            tokens = count_tokens(content)
            
            file_details.append({
                'rel_path': rel_path,
                'chars': chars,
                'tokens': tokens
            })
            
            total_files += 1
            total_chars += chars
            total_tokens += tokens
            
    # Sort files by token count descending
    file_details.sort(key=lambda x: x['tokens'], reverse=True)
    
    print(f"{'File Path':<50} | {'Characters':<12} | {'Tokens':<10}")
    print("-" * 80)
    for detail in file_details:
        print(f"{detail['rel_path']:<50} | {detail['chars']:<12,} | {detail['tokens']:<10,}")
    print("-" * 80)
    print(f"Total Files:  {total_files}")
    print(f"Total Chars:  {total_chars:,}")
    print(f"Total Tokens: {total_tokens:,}")

if __name__ == "__main__":
    main()
